server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java

```diff
@@ -0,0 +1,154 @@
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final Client client;
+    private boolean hasTriedToLoadTemplate = false;
+    private volatile boolean isEnabled = false;
+
+    public DeprecationIndexingService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, DEPRECATION_ORIGIN);
+
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Indexes a deprecation message.
+     * @param key       the key that was used to determine if this deprecation should have been be logged.
+     *                  Useful when aggregating the recorded messages.
+     * @param message   the message to log
+     * @param xOpaqueId the associated "X-Opaque-ID" header value if any, or <code>null</code>
+     * @param params    parameters to the message, if any
+     */
+    public void writeMessage(String key, String message, String xOpaqueId, Object[] params) {
+        if (this.isEnabled == false) {
+            return;
+        }
+
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("@timestamp", Instant.now().toString());
+        payload.put("key", key);
+        payload.put("message", message);
+
+        if (isNullOrEmpty(xOpaqueId) == false) {
+            payload.put("x-opaque-id", xOpaqueId);
+        }
+
+        if (params != null && params.length > 0) {
+            payload.put("params", params);
+        }
+
+        new IndexRequestBuilder(client, IndexAction.INSTANCE).setIndex(DATA_STREAM_NAME)
+            .setOpType(DocWriteRequest.OpType.CREATE)
+            .setSource(payload)
+            .execute(new ActionListener<>() {
+                @Override
+                public void onResponse(IndexResponse indexResponse) {
+                    // Nothing to do
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    LOGGER.error("Failed to index deprecation message", e);
+                }
+            });
+    }
+
+    /**
+     * Listens for changes to the cluster state, in order to know whether to toggle indexing.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        this.isEnabled = WRITE_DEPRECATION_LOGS_TO_INDEX.get(event.state().getMetadata().settings());
+
+        if (this.isEnabled == false || this.hasTriedToLoadTemplate == true) {
+            return;
+        }
+
+        // We only ever try to load the template once, because if there's a problem, we'll spam
+        // the log with the failure on every cluster state update
+        this.hasTriedToLoadTemplate = true;
+
+        if (event.state().getMetadata().templatesV2().containsKey(TEMPLATE_NAME)) {
+            return;
+        }
+
+        loadTemplate();
+    }
+
+    /*
+     * Attempts to load a template for the deprecation logs data stream
+     */
+    private void loadTemplate() {
```

Comment:
This duplicates some recent functionality in https://github.com/elastic/elasticsearch/tree/master/x-pack/plugin/stack and I believe is the only default template installed outside of x-pack, we should probably discuss where this functionality should live. cc @dakrone

Reply:
Yeah, we have a whole tool used for installing templates and policies that should be used, it's currently in xpack, but we could move it to core if needed for this (I haven't looked at the rest of this PR, just a drive-by comment)